### PR TITLE
REALMC-12640: Parse app meta config

### PR DESCRIPTION
**Pull request by NBSquare**
[REALMC-12640](https://jira.mongodb.org/browse/REALMC-12640)
### Tasks Completed
- Defined `.mdb/app_meta.json` schema
- Modified `FindApp` to search for `.mdb/app_meta.json` when looking for root directory/config files
- Skipped group ID/app ID resolve calls whenever `AppMeta` is populated by a command
- Stubbed out `/push` command.

Note: I'm not sure what to do beyond just creating the `Push` and `DoPush` methods in `import.go`. We use empty interface to pass the POST body to the endpoint so I can't validate that config version is one of those fields, and because the endpoint isn't live yet I wasn't sure how to go about testing. I'd appreciate some input on this part of the PR.

### CODE REVIEW

File: internal/local/data.go
Comment (jandersongo) [inline internal/local/data.go:28]: [q] why are these fields optional? 
I think it makes sense for the `MdbAppMeta` struct to be optional for` local.App` but what reason do we have for making the GroupID/AppID optional? 
maybe this is more a q for @christineschen

File: internal/local/app_test.go
Comment (jandersongo) [inline internal/local/app_test.go:729]: do we have tests for `local.LoadAppMeta`?

File: internal/local/app.go
Comment (jandersongo) [inline internal/local/app.go:319]: we should still be "checking for the config file" regardless of whether there is ./mdb/app_meta.json or not. 
We need to `LoadConfig`, regardless.

File: internal/local/app.go
Comment (jandersongo) [inline internal/local/app.go:297]: this loads the original config file, so it should be kept separate from this `LoadAppMeta` helper

File: internal/local/app_test.go
Comment (jandersongo) [reply internal/local/app_test.go:729]: We should also add tests for local.FindApp to check 
- app with ./mdb/app_meta.json, loads ./mdb/app_meta.json (verify contents)
- app without ./mdb/app_meta.json doesn't load anything
for both cases, we should still load the original config

File: internal/local/app.go
Comment (jandersongo) [inline internal/local/app.go:224]: similarly to WriteConfig(), you could just
`return WriteFile(filepath.Join(a.RootDir, FileMdbAppMeta.String()), 0666, bytes.NewReader(data))`
and skip the `if`

File: internal/local/app.go
Comment (NBSquare) [reply internal/local/app.go:219]: `AppMeta` is a struct so it should never be nil. I could add a check to make sure that `config_version` is not empty but as we discussed it may not matter too much and/or be the responsibility of the caller.

File: internal/commands/push/command_test.go
Comment (NBSquare) [reply internal/commands/push/command_test.go:917]: I was seeing this naming convention elsewhere in our tests. See the first test in this file for context and let me know what you think.

File: internal/commands/app/testdata/meta/hosting/metadata.json
Comment (NBSquare) [reply internal/commands/app/testdata/meta/hosting/metadata.json:1]: I honestly have no idea what this file is but I think it was created by accident. I'll delete for now.

File: internal/local/app.go
Comment (NBSquare) [reply internal/local/app.go]: Agreed on removing the call to `LoadConfig` sinc ethat's loading a separate file. But, I do think it's worth keeping the switch statement/setting `app.Config` in this file, since this information is also being "loaded in" from app_meta.

File: internal/local/app.go
Comment (jandersongo) [inline internal/local/app.go:281]: not sure how this switch statement is beneficial here. 
The config is being set right before calling `LoadConfig`

```
app := App{RootDir: path, Config: config}
if err := app.LoadConfig(); err != nil {
 	return App{}, false, err
}
```

File: internal/commands/push/command_test.go
Comment (NBSquare) [reply internal/commands/push/command_test.go:917]: Oh yeah I see that the test has a `t.Log("and should properly pass through the expected inputs")`. Not sure if that's what is going on here but I agree that this test name makes little-to-no sense so I'm just going to rename it to Jamie's suggestion.

File: internal/local/app.go
Comment (NBSquare) [reply internal/local/app.go:218]: Yep. I wrote this method when I was working on writing the app meta. The method works and has utility that I _think_ we're going to use down the line so I figured we could leave it here. If this assumption is incorrect or you don't want to merge it here, let me know... I can stash the changes for later.

File: internal/cloud/realm/import.go
Comment (makesitgo) [reply internal/cloud/realm/import.go:76]: also the scope of this ticket doesn't necessitate us creating/including the `/push` API call just yet, let's remove this for now

File: internal/local/app.go
Comment (makesitgo) [inline internal/local/app.go:271]: just a [naming/nit] ... let's remove the `Mdb` prefix from the `AppMeta` struct and file (if we're being honest to with go patterns, it probably should have been `MDB` anyways)

instead, you can add some `NameMDB`/`NameDotMDB`/`NameHiddenMDB` to `local/data.go` to help "const-ify" the `.mdb` the `".mdb"` string, which could then be included in this `filepath.Join(...)` sequence

File: internal/local/app.go
Comment (makesitgo) [reply internal/local/app.go:281]: I would actually prefer to see us avoid this here as well... `LoadAppMeta` should do just that

the logic inside `FindApp`, which should attempt to `LoadAppMeta` first before doing our iterative check of known root config files is where we'll be able to manage this stuff

File: internal/local/app.go
Comment (makesitgo) [reply internal/local/app.go:271]: yeah... this is why I'm not the biggest fan of `app_meta.json` as the filename in the first place... it really doesn't do the purpose of this file justice

however, for all intents an purposes, `local.App` is a different representation of an "app" than `realm.App` is, so I think the point stands here that `local.AppMeta` means different things too (than the `realm` counterpart)

and while I don't think we're adding any other files to `.mdb` anytime soon, that's not a reason to couple directory and filename together like that.  we only have `config.json` in the `sync/` directory but still stick with the pattern of differentiating the directories that the CLI understands and the filenames

File: internal/cloud/realm/import.go
Comment (makesitgo) [reply internal/cloud/realm/import.go:76]: how/why did this work get included here?  the JIRA ticket doesn't mention this and that endpoint isn't yet functional

Comment (NBSquare) [reply internal/cloud/realm/import.go:76]: On Friday Christine and I discussed this as part of the work we should do here. The `/push` endpoint takes a config version from the AppMeta/config, so because this seemed related to parsing `app_meta.json` it made sense to include this here.

However, it's the *command*'s responsibility to pass params to our REST APIs, and because the command and endpoint aren't live, there wasn't much I could do here beyond stubbing out the endpoint.

File: internal/local/app.go
Comment (jandersongo) [inline internal/local/app.go:308]: (haven't thoroughly tested but) what do you guys think about something more like this?
```
func resolveConfig(path string) (App, bool, error) {
	app := App{RootDir: path}
	if _, err := os.Stat(filepath.Join(path, FileMdbAppMeta.String())); err == nil {
		app.LoadAppMeta()
		switch app.AppMeta.ConfigVersion {
		case realm.AppConfigVersion20180301:
			app.Config = FileStitch
		case realm.AppConfigVersion20200603:
			app.Config = FileConfig
		case realm.AppConfigVersion20210101:
			app.Config = FileRealmConfig
		}
		return app, true, nil
	}

	for _, config := range allConfigFiles {
		_, err := os.Stat(filepath.Join(path, config.String()))
		if err != nil {
			if os.IsNotExist(err) {
				continue
			}
			return App{}, false, err
		}
		app.Config = config
		return app, true, nil
	}

	return App{}, false, nil
}

// FindApp searches upwards for the root of a Realm app project and
// returns the local app structure, a boolean indicating if the current
// working directory is part of a Realm app project, and any error that occurs
func FindApp(path string) (App, bool, error) {
	wd, wdErr := filepath.Abs(path)
	if wdErr != nil {
		return App{}, false, wdErr
	}

	for i := 0; i < maxDirectoryContainSearchDepth; i++ {
		app, found, err := resolveConfig(wd)
		if err != nil {
			return App{}, false, err
		}
		if !found {
			continue
		}
		if err := app.LoadConfig(); err != nil {
			return App{}, false, err
		}

		// if no config version is found then continue searching for the app's root directory config
		if app.ConfigVersion() == realm.AppConfigVersionZero {
			continue
		}
		return app, true, nil
		if wd == "/" {
			break
		}
		wd = filepath.Clean(filepath.Join(wd, ".."))
	}

	return App{}, false, nil
}
```

File: internal/cloud/realm/import.go
Comment (makesitgo) [reply internal/cloud/realm/import.go:76]: the `/push` endpoint will only be relevant (at least to start) for config versions > `20210101`... which we don't currently support or have a concept of in the CLI

we should be considering this `app_meta.json` file simply as a "cache" of sorts... a way for us to remember the object id versions of app and group ids.  those are id values we need when issuing _any_ admin api request from the cli.

there's going to be a separate ticket for "add support for `20220301`" in the CLI, and its at that point we should address calling the new `/push` and `/pull` endpoints

File: internal/local/app.go
Comment (NBSquare) [reply internal/local/app.go:308]: I like this approach of resolving the config. There's some error handling edge cases I think you're missing but I'll work on this approach.

**Review by jandersongo — APPROVED** [review]
LGTM!

**Review by christineschen — APPROVED** [review]
looks pretty good! thanks for going through all the back and forth (though still wait for Nick's review before merging)

File: internal/local/data.go
Comment (makesitgo) [inline internal/local/data.go:35]: [meganit] unnecessary docstrings for non-exported members (feel free to keep as is too)

**Review by makesitgo — CHANGES_REQUESTED** [review]
I'm leaving a few vague comments here so that comes with the additional comment of "feel free to reach out to me if you want to talk about any of this in more detail"... but I think we can consolidate some of this `AppMeta` vs. `ProjectInputs` logic in a few lower-level helpers (I should have totally pointed you to these helpers earlier on)

curious as to what you think about some of these things, and if there is any possibility for what i'm suggesting here!

File: internal/local/app.go
Comment (NBSquare) [reply internal/local/app.go:38]: The issue with embedding `AppMeta` is that both `AppMeta` and `AppData` have `ConfigVersion` fields which conflict, so I'll rename this to `Meta` rather than rename the meta config version field.

File: internal/cli/inputs_test.go
Comment (jandersongo) [inline internal/cli/inputs_test.go:176]: [nit] "app" --> "appID"

File: internal/cli/inputs_test.go
Comment (jandersongo) [inline internal/cli/inputs_test.go:143]: [nit]"app" --> "appID"

File: internal/commands/push/inputs_test.go
Comment (jandersongo) [inline internal/commands/push/inputs_test.go:114]: [nit] can we lowercase the "Should"s

File: internal/commands/push/inputs_test.go
Comment (jandersongo) [inline internal/commands/push/inputs_test.go:173]: "should do nothing" --> "should error"

File: internal/commands/push/inputs_test.go
Comment (jandersongo) [inline internal/commands/push/inputs_test.go:154]: what's `f`? 😛

File: internal/commands/push/inputs_test.go
Comment (jandersongo) [inline internal/commands/push/inputs_test.go:138]: "when possible" --> "when provided"

File: internal/commands/push/inputs_test.go
Comment (jandersongo) [inline internal/commands/push/inputs_test.go:161]: "if find apps cannot find app" --> "if cannot find app"

**Comment by makesitgo** [comment]
fyi @jandersongo ... I spoke with nathan offline and I'll be stealing his branch here to get this across the finish line.

I'll do my best to incorporate your latest review before I push up to my fork for another round of reviews!

**Comment by makesitgo** [comment]
closing in favor of #292

**Pull request closed without merging**
